bftengine/src/bcstatetransfer/BCStateTran.cpp

```diff
@@ -2576,114 +2656,137 @@ void BCStateTran::processData(bool lastInBatch) {
                   newBlock, newBlockIsValid, actualBlockSize, badDataFromCurrentSourceReplica, lastInBatch));
 
     if (newBlockIsValid) {
-      uint64_t firstRequiredBlock = psd_->getFirstRequiredBlock();
-      if (!lastCollectedBlockId_)
-        lastCollectedBlockId_ = firstRequiredBlock;  // TODO: find better way to initialize once
+      // TODO - fix with all other statistics
+      // uint64_t firstRequiredBlock = psd_->getFirstRequiredBlock();
+      // if (!lastCollectedBlockId_)
+      // lastCollectedBlockId_ = firstRequiredBlock;
       if (isGettingBlocks) {
         //////////////////////////////////////////////////////////////////////////
         // if we have a new block
         //////////////////////////////////////////////////////////////////////////
         ConcordAssertAND(lastChunkInRequiredBlock >= 1, actualBlockSize > 0);
 
         sourceSelector_.onReceivedValidBlockFromSource();
-        bool lastFetchedBlockIdInCycle = isLastFetchedBlockIdInCycle(nextRequiredBlock_);
-        bool minBlockIdInCurrentRange = isMinBlockIdInCurrentRange(nextRequiredBlock_);
+        bool lastFetchedBlockIdInCycle = isLastFetchedBlockIdInCycle(fetchState_.nextBlockId);
+        bool minBlockIdInCurrentBatch = fetchState_.isMinBlockId(fetchState_.nextBlockId);
 
+        // TODO - 1) report only after we put the block succesfully 2) uncomment and fix to support new protocol
         // Report collecting status for every block collected. Log entry is created every fixed window
-        // gettingMissingBlocksSummaryWindowSize. If maxBlockId is true: summarize the whole cycle without including
-        // "commit to chain duration" and vblock. In that case last window might be less than the fixed
-        // gettingMissingBlocksSummaryWindowSize.
-        reportCollectingStatus(firstRequiredBlock, actualBlockSize, lastFetchedBlockIdInCycle);
+        // gettingMissingBlocksSummaryWindowSize. If maxcommitNextMaxBlockId_BlockId is true: summarize the whole cycle
+        // without i including "commit to chain duration" and vblock. In that case last window might be less than the
+        // fixed gettingMissingBlocksSummaryWindowSize.
+        // reportCollectingStatus(firstRequiredBlock, actualBlockSize, lastFetchedBlockIdInCycle);
+
+        // WAIT_SINGLE_JOB: We have a block ready in buffer_, but no free context. let's wait for one job to finish.
+        // NO_WAIT: Opportunistic - finalize all jobs that are done, don't wait for the onging ones
+        finalizePutblockAsync(ioPool_.empty() ? PutBlockWaitPolicy::WAIT_SINGLE_JOB : PutBlockWaitPolicy::NO_WAIT);
 
-        // Put the block. We distinguishe between last block non-last block
+        // Put the block. We distinguishe between last block in cycle, last block in fetch range, and a 'regular' block
```

Comment:
should we term 'last block in fetch range' as either min/max?

Reply:
I will fix this comment, thanks.
Just to be clear:
1. I don't want to use the word last in vars. Its missleading. If I could I would almost remove all of them, except the terminology -  "last block we put in cycle".
2. In a single batch we might have multiple fetch ranges. So it is true that this is the last block in fetch range, but it would be more unique to say it is the last block in all fetch ranges - and the last block in a batch. since we don't want to use the word last, since these are block IDs, it is more clear in my opinion to use min and max, it gives a good idea, even if we decided to change again the order of block collection.